src/net_processing.cpp

```diff
@@ -947,17 +947,86 @@ void Misbehaving(NodeId pnode, int howmuch, const std::string& message) EXCLUSIV
         LogPrint(BCLog::NET, "%s: %s peer=%d (%d -> %d)%s\n", __func__, state->name, pnode, state->nMisbehavior-howmuch, state->nMisbehavior, message_prefixed);
 }
 
+/**
+ * Returns true if the given validation state result may result in us banning/disconnecting a peer
+ * which provided such an object. This is used to determine whether to relay transactions to
+ * whitelisted peers, preventing us from relaying things which would result in them disconnecting
+ * us.
+ */
 static bool MayResultInDisconnect(const CValidationState& state, bool via_compact_block) {
-    assert(!via_compact_block);
-    return (state.GetDoS() > 0);
+    switch (state.GetReason()) {
+    case ValidationInvalidReason::NONE:
+        return false;
+    // The node is providing invalid data:
+    case ValidationInvalidReason::CONSENSUS:
+    case ValidationInvalidReason::BLOCK_MUTATED:
+    case ValidationInvalidReason::CACHED_INVALID:
+        // Blocks (or headers) invalid for one of the above reasons may not be
+        // knowable to a high-bandwidth compact block peer, prior to relay.
+        // Headers that are invalid for reasons that should be known prior to
+        // block validation -- such as bad proof of work, too-early-time, or
+        // building off an invalid or missing block -- are punished regardless
+        // (see below).
+        return !via_compact_block;
+    case ValidationInvalidReason::BLOCK_INVALID_HEADER:
+    case ValidationInvalidReason::BLOCK_CHECKPOINT:
+    case ValidationInvalidReason::BLOCK_INVALID_PREV:
+    case ValidationInvalidReason::BLOCK_MISSING_PREV:
+        return true;
+    // Conflicting (but not necessarily invalid) data or different policy:
+    case ValidationInvalidReason::RECENT_CONSENSUS_CHANGE:
+    case ValidationInvalidReason::BLOCK_BAD_TIME:
+    case ValidationInvalidReason::TX_NOT_STANDARD:
+    case ValidationInvalidReason::TX_MISSING_INPUTS:
+    case ValidationInvalidReason::TX_WITNESS_MUTATED:
+    case ValidationInvalidReason::TX_CONFLICT:
+    case ValidationInvalidReason::TX_MEMPOOL_POLICY:
+        return false;
+    }
+    return false;
 }
 
+//! Returns true if the peer was punished (probably disconnected)
 static bool MaybePunishNode(NodeId nodeid, const CValidationState& state, bool via_compact_block, const std::string& message = "") {
-    int nDoS = state.GetDoS();
-    if (nDoS > 0 && !via_compact_block) {
-         LOCK(cs_main);
-         Misbehaving(nodeid, nDoS, message);
-         return true;
+    switch (state.GetReason()) {
```

Comment:
The goal of this work was to make net_processing aware of the actual reasons for validation failures, rather than just deal with opaque numbers instructing it to do something.

In the future, I'd like to make it so that we use more context to decide how to punish a peer.  One example is to differentiate inbound and outbound peer misbehaviors.  Another potential example is if we'd treat RECENT_CONSENSUS_CHANGE failures differently (ie after the next consensus change is implemented), and perhaps again we'd want to treat some peers differently than others.

I do think that the `MayResultInDisconnect` logic was not very useful, and in the latest version I've vastly simplified things to reflect that we only use this for deciding whether to relay transactions from whitelisted peers.  So there should be much less code duplication now.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/15141#discussion_r261834407

Possibly consider adding this helpful context information to the PR description.

Reply:
Done